mlfinlab/data_structures/base_bars.py

```diff
@@ -186,7 +231,334 @@ def _get_imbalance(self, price, signed_tick, volume):
+
+        # Previous bars number of ticks and previous tick imbalances
+        self.imbalance_tick_statistics = {'num_ticks_bar': [], 'imbalance_array_buy': [], 'imbalance_array_sell': [],
+                                          'buy_ticks_proportion': []}
+
+        if analyse_thresholds:
+            # Array of dicts: {'timestamp': value, 'cum_theta': value, 'exp_num_ticks': value, 'exp_imbalance': value}
+            self.bars_thresholds = []
+        else:
+            self.bars_thresholds = None
+
+        self.imbalances_are_counted_flag = False
+
+    def _reset_cache(self):
+        """
+        Implementation of abstract method _reset_cache for imbalance bars
+        """
+        self.open_price = None
+        self.high_price, self.low_price = -np.inf, np.inf
+        self.cum_statistics = {'cum_ticks': 0, 'cum_dollar_value': 0, 'cum_volume': 0, 'cum_buy_volume': 0}
+        self.thresholds['cum_theta_buy'], self.thresholds['cum_theta_sell'], self.thresholds['buy_ticks_num'] = 0, 0, 0
+
+    def _extract_bars(self, data: pd.DataFrame) -> list:
+        """
+        For loop which compiles the various run bars: dollar, volume, or tick.
+
+        :param data: (DataFrame) Contains 3 columns - date_time, price, and volume.
+        :return: (List) of bars built using the current batch.
+        """
+
+        # Iterate over rows
+        list_bars = []
+        for row in data.values:
+            # Set variables
+            date_time = row[0]
+            price = np.float(row[1])
+            volume = row[2]
+            dollar_value = price * volume
+            signed_tick = self._apply_tick_rule(price)
+
+            if self.open_price is None:
+                self.open_price = price
+
+            # Update high low prices
+            self.high_price, self.low_price = self._update_high_low(
+                self.high_price, self.low_price, price)
+
+            # Bar statistics calculations
+            self.cum_statistics['cum_ticks'] += 1
+            self.cum_statistics['cum_dollar_value'] += dollar_value
+            self.cum_statistics['cum_volume'] += volume
+            if signed_tick == 1:
+                self.cum_statistics['cum_buy_volume'] += volume
+
+            # Imbalance calculations
+            imbalance = self._get_imbalance(price, signed_tick, volume)
+
+            if imbalance > 0:
+                self.imbalance_tick_statistics['imbalance_array_buy'].append(imbalance)
+                self.thresholds['cum_theta_buy'] += imbalance
+                self.thresholds['buy_ticks_num'] += 1
+            elif imbalance < 0:
+                self.imbalance_tick_statistics['imbalance_array_sell'].append(abs(imbalance))
+                self.thresholds['cum_theta_sell'] += abs(imbalance)
+
+            self.imbalances_are_counted_flag = np.isnan([self.thresholds['exp_imbalance_buy'], self.thresholds[
+                'exp_imbalance_sell']]).any()  # Flag indicating that both buy and sell imbalances are counted
```

Comment:
It seems like a mismatch between comment and code. np.isnan will set True if value is not counted and False when it is counted. So any() will trigger True when one or more values are NOT counted.
My proposal: use (~np.isnan(...)).all(), so that we set flag as True only when both values are not nan (hence, both are counted).